Adding docs on Tracing and Jaeger for Serverless (#22090)
openshift/openshift-docs: modules/serverless-jaeger-config.adoc

// Module included in the following assemblies:
//
// * /serverless/serverless-tracing.adoc

[id="serverless-jaeger-config_{context}"]
= Configuring Jaeger for use with {ServerlessProductName}

.Procedure

. Create and apply the Jaeger custom resource:
+
----
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaeger
----

// TODO: Add images to show configuring this
. Enable tracing on for Knative Serving:
+
----

> $ oc edit ks -n knative-serving knative-serving

Reviewer: Sorry, I had a typo in original doc... Can you remove second `knative-serving`.   
```suggestion
$ oc edit ks -n knative-serving
```

----
+
.Example YAML
----
tracing:
  sample-rate: "0.1"
  backend: zipkin
  zipkin-endpoint: http://jaeger-collector.<JAEGER_NAMESPACE>.svc.cluster.local:9411/api/v2/spans
  debug: "true"
----
+
* The backend must be `zipkin`.

> // why? Check Jaeger docs.

Reviewer: It is a Knative's specific configuration. `zipikin` is has same tracing API and we can use it via Jaeger.

It is strange, but it is correct :smile_cat:

// Module included in the following assemblies:
//
// * /serverless/serverless-tracing.adoc
// */jaeger/jaeger_config/serverless-jaeger-integration.adoc

[id="serverless-jaeger-config_{context}"]
= Configuring Jaeger for use with {ServerlessProductName}

.Prerequisites

To configure Jaeger for use with {ServerlessProductName}, you will need:

* Cluster administrator permissions on an {product-title} cluster.
* A current installation of {ServerlessOperatorName} and Knative Serving.
* A current installation of Jaeger.

.Procedure

. Create and apply the Jaeger custom resource:
+
----
$ cat <<EOF | oc apply -f -
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaeger
EOF
----

. Enable tracing on for Knative Serving, by editing the namespace and adding the tracing YAML:
+
----

> $ oc edit ks -n knative-serving

Reviewer: I tested `oc apply` against `KnativeServing` resource. It seems `apply` for KnativeServing overwrites configs and so it is dangerous.

I applied this:
```
cat <<EOF | oc apply -f -
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:
    tracing:
      backend: zipkin
      zipkin-endpoint: http://jaeger-collector.default.svc.cluster.local:9411/api/v2/spans
      debug: "true"
EOF
```

Then, all configurations which are not reconciled, they are gone: https://gist.github.com/nak3/7f7092d9fa42dcbbd14e5bf85e0c0e96

> $ oc edit ks -n knative-serving

Reviewer 2: I _think_ that's expected. You would use apply only if you have the entire state of the KnativeServing local of course (like in GitOps).

All I'm saying is: This is normal operation for any Kubernetes resource. KnativeServing isn't special in any way here imo.

----
+
.Tracing YAML example
----

> tracing:

Reviewer: Yes, here is the entire context:

```yaml
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:

    ... snip ..

    tracing:
      sample-rate: "0.1"
      backend: zipkin
      zipkin-endpoint: http://jaeger-collector.<JAEGER_NAMESPACE>.svc.cluster.local:9411/api/v2/spans
      debug: "true"
```

> tracing:

Author: Awesome, thanks folks!

  sample-rate: "0.1"
  backend: zipkin
  zipkin-endpoint: http://jaeger-collector.<JAEGER_NAMESPACE>.svc.cluster.local:9411/api/v2/spans
  debug: "true"
----
+
.. The backend must be `zipkin`.
.. The `zipkin-endpoint` must point to your `jaeger-collector` service endpoint.
+
You can find this endpoint by getting the Jaeger URL:
+
----

> $ oc get route

Reviewer: Yes, it should be a cluster local address service. So, the command should be:

```
$ oc get svc
```

> $ oc get route

Reviewer 2: Even that is unnecessary right? The command you show only has a placeholder in the namespace, so the user just needs to throw the namespace they installed the Jaeger resource in in there?

> $ oc get route

Author: I made some changes to make this more readable if you want to take another look. I think I just need to know what a user would do to get that endpoint so they can use it in the YAML?

> $ oc get route

Reviewer 2: Point is: The endpoint you get here is not what you use in the YAML, but to open the Jaeger dashboard in your browser! @nak3 could you provide an example output here maybe?

----
+

> The service endpoint will be in the format `https://jaeger-<namespace>.apss.<snip>`.

Reviewer: The `https://jaeger-<namespace>.apss.<snip>` is an URL for the Jaeger's web UI. The endpoint which we want to set to `zipkin-endpoint` is different from the URL for Jaeger's web UI.

Reviewer 2: This one will need the `oc get route` command I believe? To easily fetch the URL. Example output would be great here.

Reviewer: Umm.. Then, we should write the docs like "To access Jaeger's web UI, run `oc get route` and access  to https://jaeger-<namespace>.apss.yordomain.example.com".

Reviewer 2: Right.

// Module included in the following assemblies:
//
// * /serverless/serverless-tracing.adoc
// */jaeger/jaeger_config/serverless-jaeger-integration.adoc

[id="serverless-jaeger-config_{context}"]
= Configuring Jaeger for use with {ServerlessProductName}

.Prerequisites

To configure Jaeger for use with {ServerlessProductName}, you will need:

* Cluster administrator permissions on an {product-title} cluster.
* A current installation of {ServerlessOperatorName} and Knative Serving.
* A current installation of the Jaeger Operator.

.Procedure

. Create and apply the Jaeger custom resource:
+
----
$ cat <<EOF | oc apply -f -
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaeger
  namespace: default
EOF
----

. Enable tracing for Knative Serving, by editing the `KnativeServing` resource and adding a YAML configuration for tracing.
+
.Tracing YAML example
----
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:
    tracing:

>       sample-rate: "0.1"

Reviewer 3: Might as well document the `sample-rate` ,  something like:

Defines sampling probability. With `sample-rate: "0.1"`, approximately 1 in 10 traces will be sampled.

      backend: zipkin <1>
      zipkin-endpoint: http://jaeger-collector.default.svc.cluster.local:9411/api/v2/spans <2>
      debug: "false" <3>
----
+
<1> The backend must be `zipkin`.
<2> The `zipkin-endpoint` must point to your `jaeger-collector` service endpoint. To get this endpoint, substitute the namespace where the Jaeger custom resource is applied.

> 

Reviewer: There are no `<3>` for ` debug: "false" <3>` above?

// Module included in the following assemblies:
//
// * /serverless/serverless-tracing.adoc
// */jaeger/jaeger_config/serverless-jaeger-integration.adoc

[id="serverless-jaeger-config_{context}"]
= Configuring Jaeger for use with {ServerlessProductName}

.Prerequisites

To configure Jaeger for use with {ServerlessProductName}, you will need:

* Cluster administrator permissions on an {product-title} cluster.
* A current installation of {ServerlessOperatorName} and Knative Serving.
* A current installation of the Jaeger Operator.

.Procedure

. Create and apply the Jaeger custom resource:
+
----
$ cat <<EOF | oc apply -f -
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaeger
EOF
----

. Enable tracing for Knative Serving, by editing the `KnativeServing` resource and adding a YAML configuration for tracing.
+
.Tracing YAML example
----
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:
    tracing:
      sample-rate: "0.1"
      backend: zipkin <1>
      zipkin-endpoint: http://jaeger-collector.<JAEGER_NAMESPACE>.svc.cluster.local:9411/api/v2/spans <2>

>       debug: "true"

Reviewer 3: I think we should not set the debug to true here.  (Certainly not without an explanation of what it does?)

Author: @markusthoemmes @nak3 what do you think?

Reviewer: Yes, agree. We should not set to false.

```suggestion
      debug: "false"
```

(If we add an explanation, we can note as "Enabling debug mode allows all spans to be sent to the server bypassing sampling.", which is just same explanation in [upstream](https://github.com/knative/serving/blob/a0c33cf6d3c8645118ded05b7399d428ec970650/config/core/configmaps/tracing.yaml#L52-L53) )

----
+
<1> The backend must be `zipkin`.

> <2> The `zipkin-endpoint` must point to your `jaeger-collector` service endpoint. To get this endpoint, substitute the namespace where the Jaeger Operator is installed.

Reviewer 2: This should be the namespace where the **resource** above is applied (the Jaeger resource), not the Jaeger operator.

+
.Example Jaeger service endpoint
----

> http://jaeger-collector.openshift-operators.svc.cluster.local:9411/api/v2/spans

Reviewer 2: We shouldn't use openshift-operators here as that'd hint at this being the location where the operator is installed. We could use `default` as the default namespace (as its literally the default namespace :D).

Maybe we should add the namespace to the Jaeger resource above to make that clear too?

----

.Verification  steps

> Access the Jaeger web console to see tracing data. You can access the Jaeger web console by using the `jaeger-collector` service endpoint.
> 
> . Get the `jaeger-collector` service endpoint:

Reviewer 2: ```suggestion
Access the Jaeger web console to see tracing data. You can access the Jaeger web console by using the `jaeger` route.

. Get the `jaeger` route's hostname
```

+
----
$ oc get route
----

> . Open the endpoint address in your browser. The service endpoint should be in the format `pass:c[https://jaeger-<namespace>.apps.<snip>]`.

Reviewer 2: Here's an example command:

```console
$ oc get route jaeger
NAME     HOST/PORT                         PATH   SERVICES       PORT    TERMINATION   WILDCARD
jaeger   jaeger-default.apps.example.com          jaeger-query   <all>   reencrypt     None
```

Let's change the command to be "oc get route jaeger" or "oc -n default get route jaeger" if the default namespcae is where the Jaeger resource is applied (see comments above).

[id="serverless-jaeger-config_{context}"]
= Configuring Jaeger for use with {ServerlessProductName}

.Prerequisites

To configure Jaeger for use with {ServerlessProductName}, you will need:

* Cluster administrator permissions on an {product-title} cluster.
* A current installation of {ServerlessOperatorName} and Knative Serving.
* A current installation of the Jaeger Operator.

.Procedure

. Create and apply the Jaeger custom resource:
+
----
$ cat <<EOF | oc apply -f -
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: jaeger
  namespace: default
EOF
----

. Enable tracing for Knative Serving, by editing the `KnativeServing` resource and adding a YAML configuration for tracing.
+
.Tracing YAML example
----
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:
    tracing:
      sample-rate: "0.1" <1>
      backend: zipkin <2>
      zipkin-endpoint: http://jaeger-collector.default.svc.cluster.local:9411/api/v2/spans <3>
      debug: "false" <4>
----
+
<1> The `sample-rate` defines sampling probability. Using `sample-rate: "0.1"`` means that 1 in 10 traces will be sampled.

> <2> The backend must be `zipkin`.

Reviewer 4: `backend` must be set to `zipkin`.

      debug: "true"
----
+
.. The backend must be `zipkin`.
.. The `zipkin-endpoint` must point to your `jaeger-collector` service endpoint from the previous step.

.Verification  steps

> Access the Jaeger web console to see tracing data. You can access the Jaeger web console by using the `jaeger-collector` service endpoint.

Author: There's an issue tracking docs for accessing the Jaeger console https://issues.redhat.com/browse/OSSMDOC-91 - so maybe I can leave this here for now and then switch to including that doc module once it exists? Should this be the same thing @nak3 ?

> <4> Enabling debug mode by setting `debug: "false"` allows all spans to be sent to the server, bypassing sampling.

Reviewer 3: Setting `debug: true` would enable the debug mode.  `debug: "false"` doesn't really do anything (expect pointing to users that they can enable it... )

Author: Oops, good catch, thanks @maschmid :slightly_smiling_face:

> <4> Debugging should be set to false. Enabling debug mode by setting `debug: "true"` allows all spans to be sent to the server, bypassing sampling.

Reviewer 4: s/false/`false`
